pkg/controlplane/reconcilers/lease.go

```diff
@@ -167,10 +166,7 @@ func NewLeaseEndpointReconciler(epAdapter EndpointsAdapter, masterLeases Leases)
 // different from the directory listing, and update the endpoints object
 // accordingly.
 func (r *leaseEndpointReconciler) ReconcileEndpoints(serviceName string, ip net.IP, endpointPorts []corev1.EndpointPort, reconcilePorts bool) error {
-	r.reconcilingLock.Lock()
-	defer r.reconcilingLock.Unlock()
```

Comment:
if we move this lock inside `doReconcile` we make it thread safe and avoid the race you mentioned, no?

Reply:
Maybe the safest way is keep this lock in `ReconcileEndpoints` and locked this lock in `RemoveEndpoints` meanwhile.

Reply:
> There will be no lock protection after move this lock into `doReconcile` when`UpdateLease` in `ReconcileEndpoints` and `RemoveLease` in `RemoveEndpoints`.

it is an API call, AFAIK the concurrency will be handled by etcd

Reply:
There is a possibility that call `UpdateLease` after `RemoveLease`. So user could get an unexpected error in `ReconcileEndpoints`. I looks a bit strange even though this has no side effects.

Reply:
> There is a possibility that call UpdateLease after RemoveLease

that possibility always exist, despite it is locked or not,

Reply:
yes, you're right.

Reply:
I found that `UpdateLease` will create a new key in etcd if the key doesn't exist. So I think locking the `reconcilingLock` in `RemoveEndpoints` and `ReconcileEndpoints` separately is simpler and safer than do this in `doReconcile`.
PR has updated. PTAL if you have time.

Reply:
yes, you are right, so the leases can not be updated during the reconcile block